src/pageEditor/context/connection.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2024 PixieBrix, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import type { Target } from "webext-messenger";
+import { expectContext } from "@/utils/expectContext";
+import type { Nullishable } from "@/utils/nullishUtils";
+
+// This is true enough for Page Editor purposes because we're editing the rendered frame:
+// https://developer.chrome.com/blog/extension-instantnav
+const TOP_LEVEL_FRAME_ID = 0;
+
+/**
+ * Error thrown when the Page Editor is not pointed to a tab.
+ */
+class NotInspectingTabError extends Error {
+  override name = "NotInspectingTabError";
+}
+
+function getUrlParamTabId(): Nullishable<number> {
+  const tabIdParam: Nullishable<string> = new URLSearchParams(
+    location.search,
+  ).get("tabId");
+  return tabIdParam ? Number.parseInt(tabIdParam, 10) : null;
+}
+
+let currentTabId: Nullishable<number> =
```

Comment:
I think this broadens the type unnecessarily 

```suggestion
let currentTabId =
```

Reply:
I don't think so? getUrlParamTabId will return null because URLSearchParams.get will return `null`. But the optional chained expression would return undefined if devtools or chrome is not defined

In any case, I don't think there's any value in distinguishing null vs. undefined here